chain/subscriber.go

```diff
@@ -1,29 +1,509 @@
 package chain
 
 import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
 	"time"
 
 	"go.sia.tech/core/types"
 	"go.sia.tech/coreutils/chain"
+	"go.sia.tech/coreutils/wallet"
 	"go.sia.tech/renterd/api"
+	"go.uber.org/zap"
+)
+
+const (
+	// updatesBatchSize is the maximum number of updates to fetch in a single
+	// call to the chain manager when we request updates since a given index.
+	updatesBatchSize = 1000
+)
+
+var (
+	errClosed = errors.New("subscriber closed")
 )
 
 type (
+	ChainManager interface {
+		Tip() types.ChainIndex
+		OnReorg(fn func(types.ChainIndex)) (cancel func())
+		UpdatesSince(index types.ChainIndex, max int) (rus []chain.RevertUpdate, aus []chain.ApplyUpdate, err error)
+	}
+
 	ChainStore interface {
-		BeginChainUpdateTx() (ChainUpdateTx, error)
+		ProcessChainUpdate(fn func(ChainUpdateTx) error) error
 		ChainIndex() (types.ChainIndex, error)
 	}
 
 	ChainUpdateTx interface {
-		Commit() error
-		Rollback() error
-
 		ContractState(fcid types.FileContractID) (api.ContractState, error)
 		UpdateChainIndex(index types.ChainIndex) error
 		UpdateContract(fcid types.FileContractID, revisionHeight, revisionNumber, size uint64) error
 		UpdateContractState(fcid types.FileContractID, state api.ContractState) error
 		UpdateContractProofHeight(fcid types.FileContractID, proofHeight uint64) error
 		UpdateFailedContracts(blockHeight uint64) error
 		UpdateHost(hk types.PublicKey, ha chain.HostAnnouncement, bh uint64, blockID types.BlockID, ts time.Time) error
+
+		wallet.UpdateTx
+	}
+
+	ContractStore interface {
+		AddContractStoreSubscriber(context.Context, ContractStoreSubscriber) (map[types.FileContractID]struct{}, func(), error)
+	}
+
+	ContractStoreSubscriber interface {
+		AddContractID(fcid types.FileContractID)
+	}
+
+	Subscriber struct {
+		cm              ChainManager
+		cs              ChainStore
+		csUnsubscribeFn func()
+		logger          *zap.SugaredLogger
+
+		announcementMaxAge time.Duration
+		retryTxIntervals   []time.Duration
```

Comment:
Why not leave this up to the store? The store interface will also know if it failed due to a lock so you don't retry transactions that have no chance of succeeding. 

I'd also probably use a backoff function with a maximum interval instead of hardcoding the intervals.

Reply:
Using the callback made this possible, didn't spot I could remove these retry intervals, thanks!